Simulation and Modeling Lab  
**Course Code:** CSE 413  
**Course Teacher:** Afroja Ahmed Smrity, Lecturer, Department of Computer Science and Engineering, University of Information Technology and Sciences (UITS)  
**Submitted By:** Sumaia Bintey Ismail  
**ID:** 0432320005101103  
**Submission Date:** 10 August, 2026  


# Necessary Libraries

In [17]:
import numpy as np
from sympy import Matrix
import matplotlib.pyplot as plt
import seaborn as sns

#Task 1. Create Your Personal Matrix  
**Let:**  
● d1 = last digit of your student ID  
● d2 = second last digit of your student ID
Create the matrix:  
  A =   [d1+2  d2+1]
        [2d1  d2+2]  

In [18]:
id = "0432320005101103"
d1 = int(id[-1])
d2 = int(id[-2])
print(f"Student ID: {id}")
print(f"d1 (last digit): {d1}")
print(f"d2 (second last digit): {d2}\n")
A = np.array([
    [d1 + 2, d2 + 1],
    [2 * d1, d2 + 2]
])
print("--- Matrix A ---")
print(A)

Student ID: 0432320005101103
d1 (last digit): 3
d2 (second last digit): 0

--- Matrix A ---
[[5 1]
 [6 2]]


# Task 2. Analyze Matrix A
**Using NumPy, compute:**   
● Shape of A    
● Determinant   
● Rank    
● Eigenvalues   
● Inverse (only if determinant ≠ 0)

In [19]:
shape_A = A.shape
det_A = np.linalg.det(A)
rank_A = np.linalg.matrix_rank(A)
eigenvalues_A = np.linalg.eigvals(A)
print("Properties of Matrix A:")
print(f"Shape: {shape_A}")
print(f"Determinant: {det_A:.4f}")
print(f"Rank: {rank_A}")
print(f"Eigenvalues: {eigenvalues_A}")
if abs(det_A) > 1e-9:
    inv_A = np.linalg.inv(A)
    print(f"Inverse of A:\n{inv_A}")
else:
    print("Inverse of A: Does not exist (Matrix is singular).")


Properties of Matrix A:
Shape: (2, 2)
Determinant: 4.0000
Rank: 2
Eigenvalues: [6.37228132 0.62771868]
Inverse of A:
[[ 0.5  -0.25]
 [-1.5   1.25]]


# Task 3. Value Change Experiment   
**Create a new matrix B by changing exactly one value in A:**   
● Add +1 to one element of your choice    
● You must clearly state which element you changed



In [20]:
B = A.copy()
B[0, 0] += 1
print("--- Matrix B ---")
print("(Created by adding +1 to the top-left element [0,0])")
print(B)

--- Matrix B ---
(Created by adding +1 to the top-left element [0,0])
[[6 1]
 [6 2]]


# Task 4. Re-analyze Matrix B
**Compute the same properties as in Step 2.**

In [21]:
shape_B = B.shape
det_B = np.linalg.det(B)
rank_B = np.linalg.matrix_rank(B)
eigenvalues_B = np.linalg.eigvals(B)
print("Properties of Matrix B:")
print(f"Shape: {shape_B}")
print(f"Determinant: {det_B:.4f}")
print(f"Rank: {rank_B}")
print(f"Eigenvalues: {eigenvalues_B}")
if abs(det_B) > 1e-9:
    inv_B = np.linalg.inv(B)
    print(f"Inverse of B:\n{inv_B}")
else:
    print("Inverse of B: Does not exist (Matrix is singular).")

Properties of Matrix B:
Shape: (2, 2)
Determinant: 6.0000
Rank: 2
Eigenvalues: [7.16227766 0.83772234]
Inverse of B:
[[ 0.33333333 -0.16666667]
 [-1.          1.        ]]


# Task 5. Answer the following in your own words:
**1. How did the determinant change and why?**  
The determinant changed from 4.0 to 6.0.
Mathematical Reason: The determinant of a $2 \times 2$ matrix $M = \begin{bmatrix} a & b \\ c & d \end{bmatrix}$ is calculated as $\det(M) = ad - bc$. By adding $+1$ to the top-left element $a$, the new matrix becomes $M' = \begin{bmatrix} a+1 & b \\ c & d \end{bmatrix}$. The new determinant is $(a+1)d - bc = ad + d - bc = \det(M) + d$. Therefore, the determinant predictably increased by exactly the value of the bottom-right element $d$, which in my matrix is 2.  
**2. Did the rank change? Explain.**  
The rank remained the same at 2.
Reason: The rank of a $2 \times 2$ matrix indicates the number of linearly independent rows or columns. As long as the determinant is non-zero, the matrix remains full rank. Because changing one element by $+1$ changed the determinant from 4 to 6 (neither of which are 0), the rows remained linearly independent, preserving the rank.   
**3. How did the eigenvalues respond to the value change?**   
The eigenvalues shifted from approximately [6.37, 0.63] in Matrix A to [7.16, 0.84] in Matrix B.

Reason: Eigenvalues are dictated by the Trace (sum of the main diagonal) and the Determinant of a matrix via the characteristic equation $$det(A - \lambda I) = 0$$. By adding $+1$ to the top-left element, I increased the Trace of the matrix from 7 to 8, and the determinant increased from 4 to 6. The eigenvalues had to shift outward to satisfy these newly increased parameters.   
**4. Is B easier or harder to invert than A? Why?**   
Computationally, Matrix B is slightly harder (less stable) to invert than Matrix A. The ease of inverting a matrix is heavily tied to how close its determinant is to zero. Matrices with determinants very close to 0 are highly sensitive to floating-point errors (they have a high "condition number"). Since Matrix B's determinant (4.0) moved closer to 0 compared to Matrix A's determinant (6.0), Matrix B is inching closer to being a singular (non-invertible) matrix.
